qiita_db/handlers/oauth2.py

```diff
@@ -0,0 +1,306 @@
+# -----------------------------------------------------------------------------
+# Copyright (c) 2014--, The Qiita Development Team.
+#
+# Distributed under the terms of the BSD 3-clause License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# -----------------------------------------------------------------------------
+
+from base64 import urlsafe_b64decode
+from string import ascii_letters, digits
+import datetime
+from random import SystemRandom
+import functools
+from traceback import format_exception
+
+from tornado.web import RequestHandler
+
+from moi import r_client
+from qiita_core.exceptions import (IncorrectPasswordError, IncorrectEmailError,
+                                   UnverifiedEmailError)
+import qiita_db as qdb
+
+
+def login_client(client_id, client_secret=None):
+    """Login client_id, optionally with client_secret
+
+    Parameters
+    ----------
+    client_id : str
+        The client ID to validate
+    client_secret : str, optional
```

Comment:
`client_secret` as optional is so it can also be used to check client_id for the `Resource Owner Password Credentials` login type, which only requires client ID and is in the database as a client ID with a NULL client secret.